File: Matrix/36. Valid Sudoku.py
"""36. Valid Sudoku"""


class Solution:
    def isValidSudoku(self, board: list[list[str]]) -> bool:
        rows: dict[int, set[str]] = {}
        cols: dict[int, set[str]] = {}
        squares: dict[int, set[str]] = {}

        for i, row in enumerate(board) :
            for j, value in enumerate(row):
                if value == ".":
                    continue

                s = (i // 3) * 3 + j // 3
                
                if i not in rows:
                        rows[i] = set()
                if j not in cols:
                        cols[j] = set()
                if s not in squares:
                        squares[s] = set()
                
                if value in rows[i] or value in cols[j] or value in squares[s]:
                    return False
                rows[i].add(value)
                cols[j].add(value)
                squares[s].add(value)

        return True


def main() -> None:
    """Main Function"""
    board = [["5", "3", ".", ".", "7", ".", ".", ".", "."],
             ["6", ".", ".", "1", "9", "5", ".", ".", "."],
             [".", "9", "8", ".", ".", ".", ".", "6", "."],
             ["8", ".", ".", ".", "6", ".", ".", ".", "3"],
             ["4", ".", ".", "8", ".", "3", ".", ".", "1"],
             ["7", ".", ".", ".", "2", ".", ".", ".", "6"],
             [".", "6", ".", ".", ".", ".", "2", "8", "."],
             [".", ".", ".", "4", "1", "9", ".", ".", "5"],
             [".", ".", ".", ".", "8", ".", ".", "7", "9"]]

    print(Solution().isValidSudoku(board))


if __name__ == "__main__":
    main()

File: Matrix/36. Valid Sudoku.cpp
#include <bits/stdc++.h>
using namespace std;

class Solution {
public:
    bool isValidSudoku(vector<vector<char>>& board) {
        unordered_map<int, unordered_set<char>> rows;
        unordered_map<int, unordered_set<char>> cols;
        unordered_map<int, unordered_set<char>> squares;        

        for (int i = 0; i < 9; i++) {
            for (int j = 0; j < 9; j++) {
                if (board[i][j] == '.' ) continue;

                int squareIndex = (i / 3) * 3 + j / 3;
                if (rows[i].find(board[i][j]) != rows[i].end() || 
                    cols[j].find(board[i][j]) != cols[j].end() || 
                    squares[squareIndex].find(board[i][j]) != squares[squareIndex].end()) return false;

                rows[i].insert(board[i][j]);
                cols[j].insert(board[i][j]);
                squares[squareIndex].insert(board[i][j]);
            }
        }

        return true;
    }
};


int main() {
    vector<vector<char>> board = {
        {'5','3','.','.','7','.','.','.','.'},
        {'6','.','.','1','9','5','.','.','.'},
        {'.','9','8','.','.','.','.','6','.'},
        {'8','.','.','.','6','.','.','.','3'},
        {'4','.','.','8','.','3','.','.','1'},
        {'7','.','.','.','2','.','.','.','6'},
        {'.','6','.','.','.','.','2','8','.'},
        {'.','.','.','4','1','9','.','.','5'},
        {'.','.','.','.','8','.','.','7','9'}
    };

    Solution sol;
    if (sol.isValidSudoku(board) == true) {
        cout << "True";
    } else {
        cout << "False";
    }

    return 0;
}